sentry_sdk/integrations/starlette.py

```diff
@@ -0,0 +1,459 @@
+
+
+def _capture_exception(exception, handled=False):
+    # type: (BaseException, **Any) -> None
+    hub = Hub.current
+    if hub.get_integration(StarletteIntegration) is None:
+        return
+
+    event, hint = event_from_exception(
+        exception,
+        client_options=hub.client.options if hub.client else None,
+        mechanism={"type": StarletteIntegration.identifier, "handled": handled},
+    )
+
+    hub.capture_event(event, hint=hint)
+
+
+def patch_exception_middleware(middleware_class):
+    # type: (Any) -> None
+    """
+    Capture all exceptions in Starlette app and
+    also extract user information.
+    """
+    old_middleware_init = middleware_class.__init__
+
+    def _sentry_middleware_init(self, *args, **kwargs):
+        # type: (Any, Any, Any) -> None
+        old_middleware_init(self, *args, **kwargs)
+
+        # Patch existing exception handlers
+        for key in self._exception_handlers.keys():
+            old_handler = self._exception_handlers.get(key)
+
+            def _sentry_patched_exception_handler(self, *args, **kwargs):
+                # type: (Any, Any, Any) -> None
+                exp = args[0]
+                _capture_exception(exp, handled=True)
+                return old_handler(self, *args, **kwargs)
+
+            self._exception_handlers[key] = _sentry_patched_exception_handler
+
+    middleware_class.__init__ = _sentry_middleware_init
+
+    old_call = middleware_class.__call__
+
+    async def _sentry_exceptionmiddleware_call(self, scope, receive, send):
+        # type: (Dict[str, Any], Dict[str, Any], Callable[[], Awaitable[Dict[str, Any]]], Callable[[Dict[str, Any]], Awaitable[None]]) -> None
+        # Also add the user (that was eventually set by be Authentication middle
+        # that was called before this middleware). This is done because the authentication
+        # middleware sets the user in the scope and then (in the same function)
+        # calls this exception middelware. In case there is no exception (or no handler
+        # for the type of exception occuring) then the exception bubbles up and setting the
+        # user information into the sentry scope is done in auth middleware and the
+        # ASGI middleware will then send everything to Sentry and this is fine.
+        # But if there is an exception happening that the exception middleware here
+        # has a handler for, it will send the exception directly to Sentry, so we need
+        # the user information right now.
+        # This is why we do it here.
+        _add_user_to_sentry_scope(scope)
+        await old_call(self, scope, receive, send)
+
+    middleware_class.__call__ = _sentry_exceptionmiddleware_call
+
+
+def _add_user_to_sentry_scope(scope):
+    # type: (Dict[str, Any]) -> None
+    """
+    Extracts user information from the ASGI scope and
+    adds it to Sentry's scope.
+    """
+    if "user" not in scope:
+        return
+
+    if not _should_send_default_pii():
+        return
```

Comment:
This is great for Starlette!

For FastAPI, the `AuthenticationMiddleware` is probably not commonly used because that is handled with dependencies, and there's no predefined expected/standardized user data/model.

It would probably make sense to explain in the docs how to manually and optionally record user data for FastAPI users in dependencies.

Reply:
Good to know. What is the most used dependency for user handling?

We have something in the documentation on how to set the user: https://docs.sentry.io/platforms/python/enriching-events/identify-user/
I will make sure it lands in the FastAPI docs too!

Reply:
That sounds perfect!

---

There's not an expected predefined user data model or dependency, so every application can use the data model they find more useful. It normally doesn't require adopting a predefined model or any type of dependency, because installing, learning, and using some package would require the same amount of effort and code as implementing it directly.

The documentation explaining how to handle users is here: https://fastapi.tiangolo.com/tutorial/security/get-current-user/

But again, it's just explaining the concepts, not enforcing a data model.

---

I think that documenting how to handle PII in the developers' code, as you suggest with the existing docs, would probably be the best approach here too. Instead of trying to solve it behind the scenes as there's nothing obvious to expect.